### PR TITLE
[WIP] Pre backward hooks for nn.Module

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #63072
* __->__ #63071

Adds a pre-backward hook for `nn.Module` as described in https://github.com/pytorch/pytorch/issues/62998. The main use case for now is to enable profiling of module-level backwards pass.

Differential Revision: [D30240465](https://our.internmc.facebook.com/intern/diff/D30240465/)

### CODE REVIEW

File: torch/nn/modules/module.py [hunk 1]
```diff
@@ -1089,6 +1095,17 @@ def _call_impl(self, *input, **kwargs):
...
             result = bw_hook.setup_output_hook(result)
 
+        if self._pre_backward_hooks:
+            var = result
+            while not isinstance(var, torch.Tensor):
+                if isinstance(var, dict):
+                    var = next((v for v in var.values() if isinstance(v, torch.Tensor)))
+                else:
+                    var = var[0]
+            grad_fn = var.grad_fn
```
Comment (jbschlosser) [inline torch/nn/modules/module.py:1105]: nit: `grad_fn` is unused - should check it's not None before registering the hook

File: test/test_autograd.py [hunk 2]
```diff
@@ -101,6 +101,29 @@ def _function_test(self, cls):
 
         return x, y
 
+    def test_bwd_pre_hook(self):
```
Comment (jbschlosser) [inline test/test_autograd.py:104]: I think there should be another test that verifies returning a new gradient from a pre-backwards hook

**Review by jbschlosser — COMMENTED** [review]
Thanks for the addition! General approach looks good to me- some docs & a bit more testing will finish it out nicely :) Private API to start out seems good to me.

Note: I think you'll need some logic in `Module.__setstate__` to set `_pre_backward_hooks` to empty for backwards compatibility.

Also, I think we will need the global version of these pre-backwards hooks as well - cc @albanD to confirm

File: torch/nn/modules/module.py [hunk 3]
```diff
@@ -1089,6 +1095,17 @@ def _call_impl(self, *input, **kwargs):
         if bw_hook:
             result = bw_hook.setup_output_hook(result)
 
+        if self._pre_backward_hooks:
+            var = result
+            while not isinstance(var, torch.Tensor):
```
Comment (albanD) [inline torch/nn/modules/module.py:1100]: This is not the right way to do backward hook. This might trigger at the wrong time and not contain all gradients. Like the old backward hooks.
You want to do something closer to the full backward hooks above.

**Review by albanD — COMMENTED** [review]
Sorry for the late review. I wrote one but it never got sent..

**Comment by albanD** [comment]
We should do this with full hooks and not regular hooks if we really want it.
